### PR TITLE
keep @/, very magic, search in visual

### PR DESCRIPTION
**Pull request by seletskiy**
Proposed pull request will do following things:
* `@/` will be filled with successfull search pattern in search/replace modes;
* option added that allows user to user `very magic` patterns (`\v`) without specifying it in the pattern, so autocompletion will work;
* **search and substitute modes can now highlight matches only in visual selection** (:+1: for highlighting visual selected text);
* while using vim-over for searching (`:OverCommandLine /`), match that will be used first highlighted in the same way as in https://github.com/dahu/SearchParty plugin;
* vim-over plays nice with https://github.com/dahu/SearchParty plugin.

### CODE REVIEW

**Comment by osyo-manga** [comment]
Thanks for pull request.
But, merge is difficult because the change is large

**Comment by seletskiy** [comment]
@osyo-manga: there is a gif: 
a

**Comment by osyo-manga** [comment]
Please pull request for each function.

**Comment by osyo-manga** [comment]
> search and substitute modes can now highlight matches only in visual selection (:+1: for highlighting visual selected text);

This is a good pull request.

But,

> @/ will be filled with successfull search pattern in search/replace modes;
> option added that allows user to user very magic patterns (\v) without specifying it in the pattern, so autocompletion will work;

It is difficult to merge.

**Comment by seletskiy** [comment]
@osyo-manga: come on, it's just a 70 lines change. `very-magic` and visual highlighing can't be splited, because of special literals used for matching regexp in visual are different in `very-magic` and `no-magic` modes.

`keep_search` options (`@/`) introduces only 3 new lines, it's will be more bookkeeping to split it into different PR.

**Comment by seletskiy** [comment]
@osyo-manga: I've got rid off `very_magic` stuff, because I found one bug and it's very difficult to fix. Everything else remained the same. So, I guess, autocompletion should be smart enough to complete words even there is `\v` in the beginning of the line.

**Comment by seletskiy** [comment]
@osyo-manga: ping.

**Comment by osyo-manga** [comment]
Sorry, I'm busy :(

>So, I guess, autocompletion should be smart enough to complete words even there is \v in the beginning of the line.

Could we discuss about that in another new issue?

File: autoload/over/command_line/search.vim [hunk 1]
```diff
@@ -16,29 +17,60 @@ function! s:search_hl_off()
...
+			call s:search_hl_on((&ignorecase ? '\c' : "") . s:pattern)
 		endif
 		if g:over#command_line#search#enable_move_cursor
+			call setpos(".", s:old_pos)
 			if line =~ '^/.\+'
-				silent! call search(pattern, "c")
+				silent! call searchpos(s:pattern, "")
 			else
-				silent! call search(pattern, "cb")
+				silent! call searchpos(s:pattern, "b")
```
Comment (justinmk) [inline autoload/over/command_line/search.vim]: Why the change to `searchpos`?

File: autoload/over/command_line/search.vim [hunk 2]
```diff
@@ -16,29 +17,60 @@ function! s:search_hl_off()
 		call matchdelete(s:search_id)
 		unlet s:search_id
 	endif
+	if exists("s:current_search_id") && s:current_search_id != -1
+		call matchdelete(s:current_search_id)
+		unlet s:current_search_id
+	endif
+	if exists("s:cursor_search_id") && s:cursor_search_id != -1
```
Comment (justinmk) [inline autoload/over/command_line/search.vim]: This line is equivalent to:

    if get('s:', 'cursor_search_id', -1) != -1

File: autoload/over/command_line/substitute.vim [hunk 3]
```diff
@@ -105,7 +105,7 @@ endfunction
 
 
 function! s:undo()
-	if s:undo_flag
+	if exists('s:undo_flag') && s:undo_flag
```
Comment (justinmk) [inline autoload/over/command_line/substitute.vim]: This line is equivalent to:

    if get('s:', 'undo_flag', 0)

File: autoload/over/command_line/search.vim (hunk 1 above)
Comment (seletskiy) [inline autoload/over/command_line/search.vim]: Because it's expected behaviour when searching. I expect cursor to be moved on the first search result, not just highlight the matches. Like usual searching through `/` works.

File: autoload/over/command_line/substitute.vim (hunk 3 above)
Comment (seletskiy) [inline autoload/over/command_line/substitute.vim]: Actually, should be `if get(s:, 'undo_flag', 0)`

File: autoload/over/command_line/search.vim (hunk 1 above)
Comment (justinmk) [inline autoload/over/command_line/search.vim]: search() moves the cursor as long as 'n' flag is not present. The purpose of `searchpos()` is that it returns the line and column. But you aren't using the return value...

Comment (seletskiy) [inline autoload/over/command_line/search.vim]: Oh, sorry, looks like I was confused about `searchpos` call while beating with one of the bugs. Reverting...

**Comment by kovetskiy** [comment]
@osyo-manga ping?

**Comment by evanrelf** [comment]
@seletskiy I'm going to use your fork until @osyo-manga accepts your pull request :+1:

**Comment by kovetskiy** [comment]
@evanrelf: :+1:

**Comment by osyo-manga** [comment]
I created new issues.

**Comment by balta2ar** [comment]
I love the activity in this pull request! Can't wait for the features to be merged! Thanks!

**Comment by osyo-manga** [comment]
> I love the activity in this pull request! Can't wait for the features to be merged! Thanks!

Sorry, I can't merge for pull request.
Because change is large.
But, I'll implement this.

**Comment by kovetskiy** [comment]
@osyo-manga  seriously?
```
73 additions and 9 deletions. 
```

**Comment by haya14busa** [comment]
I guess he means this PR includes more than one feature which can be separated into each one PR.
The number of line changes doesn't matter.

**Comment by kovetskiy** [comment]
@haya14busa so, it is such difficult to merge?

**Comment by haya14busa** [comment]
The problem is not the difficulty but the development style I guess.
For example, if there were bugs in this PR, it's hard to detect the bad commits and difficult to revert it.
However, if each feature is separatory pull-requested, it's easy to see the diff for each feature and reverting the commit is really easy.

**Comment by kovetskiy** [comment]
@haya14busa I agree about commits, but separating features to another pull-requests, in this case, is bureaucracy.

**Comment by haya14busa** [comment]
I don't know why you think it's bureaucracy but I don't care.
The problem is @osyo-manga doesn't think so. (and me too.)
I think it's better to just wait for his implementation instead of ping him to merge this PR if @seletskiy doesn't plan to separate this PR.

**Comment by seletskiy** [comment]
@haya14busa: I see no point of separation if @osyo-manga will going to re-implement it. Just hope that his solution will be at least not worse, than mine.

**Comment by balta2ar** [comment]
@osyo-manga Why re-implement the same functionality? Splitting into different PRs is easier and faster.
@seletskiy Could you be so kind to split this PR, please? Will this work for you, @osyo-manga?
